### PR TITLE
Add --allow-self-paid and --allow-payments flags to splitlunch refresh

### PR DESCRIPTION
**Pull request by ebroder**
# Description

This introduces two new flags to `splitlunch refresh` - `--allow-self-paid` and `--allow-payments`, which override the historical behavior of filtering out those classes of transactions.

In order for this logic to behave correctly, I had to somewhat rework how splitlunch calculates the financial impact of a transaction. To convince myself I wasn't making things worse, I attempted to make it easier to test the logic around that calculation and added some tests using transactions from my Splitwise history (though I scrubbed the personal details of the other individuals involved).

For review purposes, I did my best to separate the refactoring into distinct steps which hopefully each make sense on their own; as such, I think this PR may be easiest to review commit-by-commit. (The commit messages additionally include some explanation of why I think each change represents an improvement)

As a final note, this change does lightly conflict with #71 - they both introduce flags to the `refresh_splitwise_transactions` function and make nearby edits. The conflicts should be fairly straightforward to resolve, but I'm happy to do whatever makes this easiest to manage. (I considered adjusting #71 to include these changes, but felt that it was better to leave that change simple and easy to review, given the meatiness of this change)

# Has This Been Tested?

In addition to the new automated testing around the calculations to financial impact, I've been testing with `poetry run lunchable plugins splitlunch refresh --dated-after 2023-01-01 --allow-self-paid --allow-payments`, using a local branch which contains both this change and #71. My Splitwise history includes several classes of transactions which were previously handled incorrectly, and I've verified that they are now being created with the right data.

# Checklist:

- [x] I've read the [contributing guidelines](https://juftin.com/lunchable/contributing) of this project
- [x] I've installed and used `.pre_commit` on all my code
- [x] I have documented my code, particularly in hard-to-understand areas
- [x] I have made any necessary corresponding changes to the documentation

### CODE REVIEW

**Comment by juftin** [comment]
Hey @ebroder just starting to look into PRs  - I've had some personal stuff going on and haven't been able to get around to it. Thank you in the meantime for all the detailed notes.

**Comment by juftin** [comment]
Would you bring in the latest changes from `main` - there are two commits (one of which is a CI/CD bugfix for forked repositories)

File: lunchable/plugins/splitlunch/lunchmoney_splitwise.py [hunk 1]
```diff
@@ -1080,12 +1045,17 @@ def filter_relevant_splitwise_expenses(
...
+        3) If the --allow-self-paid flag is not provided, it filters out `self-paid` expenses. A
+        `self-paid` expense is an expense in Splitwise where you originated the payment. This is
+        excluded because it is assumed that these transactions will have already been imported via a
+        different account.
 
-        3) It filters out payments. Payments are excluded because it is assumed that these
-        transactions will have already been imported via a different account.
+        4) If the --allow-payments flag is not provided, it filters out payments. Payments are
+        excluded because it is assumed that these transactions will have already been imported via a
+        different account.
```
Comment (juftin) [inline lunchable/plugins/splitlunch/lunchmoney_splitwise.py:1058]: The content in these docstrings is great - is there any of this that's worth surfacing in the public facing plugin docs @ [docs/source/splitlunch.rst](https://github.com/juftin/lunchable/blob/main/docs/source/splitlunch.rst). (I know, it's an `RST` and not `MD` 😞)?

**Comment by juftin** [comment]
This all looks really solid, the refactor is large but the feature flags look to preserve all the original functionality - the commit messages were great for understanding what's going on. Admittedly it's difficult for me to test this with my real-world data because I stopped using Splitwise with my partner a while back and also stopped using it alongside LunchMoney. I'm confident with all of these changes though and am good to move this work forward. 

I'm certainly not signing you up for any work at all @ebroder so please don't feel obligated but given that I've stopped using Splitwise alongside LunchMoney - you may be the most knowledgeable person on SplitLunch moving forward. Any updating of the docs (beyond the great docstrings) to reflect these new changes would be greatly appreciated. Thank you again for these awesome contributions.

File: lunchable/plugins/splitlunch/lunchmoney_splitwise.py (hunk 1 above)
Comment (ebroder) [reply lunchable/plugins/splitlunch/lunchmoney_splitwise.py:1058]: Sure, seems reasonable. Added some stuff.

**Comment by ebroder** [comment]
> Hey @ebroder just starting to look into PRs - I've had some personal stuff going on and haven't been able to get around to it. Thank you in the meantime for all the detailed notes - excited for these new features.

Don't worry about it! Life happens.

> Would you bring in the latest changes from main - there are two commits (one of which is a CI/CD bugfix for forked repositories)

Done.

> I'm certainly not signing you up for any work at all @ebroder so please don't feel obligated but given that I've stopped using Splitwise alongside LunchMoney - you may be the most knowledgeable person on SplitLunch moving forward. Any updating of the docs (beyond the great docstrings) to reflect these new changes would be greatly appreciated. Thank you again for these awesome contributions.

Haha, I'm tentatively OK with that. Feel free to bounce stuff my way, although I'll note that I'm primarily interested in the auto-importer and am less confident in how the Lunch Money => Splitwise stuff works.

**Comment by juftin** [comment]
Hey @ebroder apologies for the delay on this - I just added one last commit to resolve the linting errors. I've also added a commit with the necessary emojis to trigger a release. One last question for you, do you prefer to squash merge or rebase merge this branch?

**Comment by ebroder** [comment]
I think a squash merge is fine. There's some potentially interesting stuff in the history, but most of it is probably only useful for the initial code review, and it'll always be here if we need it.

**Pull request merged**